ROS2 implementation for human_model_generation module

charsyme opened the pull request.

charsyme commented:
I added the "opendr_interfaces" package for a custom srv (Mesh.srv), since I couldn't implement it in an existing package (opendr_simulation or opendr_ros2_bridge) in ROS2.

charsyme commented:
This PR is ready for review.

tsampazk commented:
Hey @charsyme! I resolved the conflicts for this PR. Apart from a conflict in bridge imports, the other two were in the to/from pose methods and were caused by changes i did on the base branch. 

I also created a [`opendr_ros2_messages` CMake package](https://github.com/opendr-eu/opendr/pull/256/commits/49f6e58be0c2cc1e383c8f353bfff8e0ae27cae7) in the base branch, and i noticed that you created a similar `opendr_interfaces` CMake package.

Do you think we could merge the messages package i created into your interfaces package that will contain all custom msgs ans srvs, as it seems more general? @passalis what do you think?

passalis commented:
> Do you think we could merge the messages package i created into your interfaces package that will contain all custom msgs ans srvs, as it seems more general? @passalis what do you think?

Sounds good, since this would ensure that we have all custom msgs and srvs in the same place.

tsampazk reviewed (changes requested):
Hey @charsyme, thank you for the new package/nodes! I've left a review with some comments to improve the package a bit and bring it up to speed with the rest of the ROS2 workspace before merging.


I tried to run the nodes but got the following error. Got any ideas on what goes wrong?
```
File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 118, in await_or_execute
    return callback(*args)
  File "~/opendr/projects/opendr_ws_2/install/opendr_simulation/lib/python3.8/site-packages/opendr_simulation/human_model_generation_service.py", line 56, in gen_callback
    output = self.model_generator.infer([rgb_img], [msk_img], extract_pose=True)
  File "~/opendr/src/opendr/simulation/human_model_generation/pifu_generator_learner.py", line 91, in infer
    raise e
  File "~/opendr/src/opendr/simulation/human_model_generation/pifu_generator_learner.py", line 80, in infer
    [verts, faces, colors] = self.evaluator.eval(self.evaluator.load_image(imgs_rgb[0], imgs_msk[0]), use_octree=True)
TypeError: cannot unpack non-iterable NoneType object
```

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+        colors = np.zeros([len(ros_colors), 3])
+        for i in range(len(ros_colors)):
+            colors[i] = np.array([ros_colors[i].r, ros_colors[i].g, ros_colors[i].b])
+        return colors
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+            keypoint.results[0].hypothesis.class_id = str(pose.id)
+            keypoint.results[0].hypothesis.score = 1.0
+            keypoints.detections.append(keypoint)
+        return keypoints
```
Done

charsyme replied:
Done

charsyme commented on projects/opendr_ws_2/src/opendr_interfaces/package.xml:
```diff
@@ -0,0 +1,27 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>opendr_interfaces</name>
+  <version>0.0.0</version>
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_interfaces/package.xml:
```diff
@@ -0,0 +1,27 @@
+
+<exec_depend>rosidl_default_runtime</exec_depend>
+
+<member_of_group>rosidl_interface_packages</member_of_group>
```
opendr_interfaces has been removed. Its content is merged with ioendr_ros2_messages.

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
         return bounding_box_list
 
-    def to_ros_face(self, category):
+    def from_ros_mesh(self, mesh_ROS, vertex_colors_ROS=None):
```
I removed it.

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+    def to_ros_colors(self, colors):
+        """
+        Converts an array of vertex_colors to a list of ROS colors
+        :type colors: numpy array (Nx3)
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+            ros_colors.append(color)
+        return ros_colors
+
+    def from_ros_3Dpose(self, ros_pose):
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/package.xml:
```diff
@@ -0,0 +1,37 @@
+  <description>OpenDR ROS2 nodes for the simulation package</description>
+  <maintainer email="tefas@csd.auth.gr">OpenDR Project Coordinator</maintainer>
+  <license>Apache License v2.0</license>
+  <depend>rclcpp></depend>
```
I removed rclcpp.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,74 @@
+from opendr.simulation.human_model_generation.utilities.model_3D import Model_3D
+
+
+class Human_model_generation_client(Node):
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
+
+
+class Pifu_service(Node):
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+    except:
+        device = 'cpu'
+    rclpy.init()
+    pifu_service = Pifu_service(device=device)
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Creates a ROS Service for human model generation
+        :param 'human_model_generation': The name of the service
+        :type input_image_topic: str
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+        """
+        Converts an array of vertex_colors to a list of ROS colors
+        :type colors: numpy array (Nx3)
+        :return colors: a list of the colors of the vertices
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+    pifu_service = Pifu_service(device=device)
+    rclpy.spin(pifu_service)
+    rclpy.shutdown()
+
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,74 @@
+    cv2.imwrite('./rendering.png', out_imgs[0].opencv())
+    client.destroy_node()
+    rclpy.shutdown()
+
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,74 @@
+    if extract_pose:
+        [out_imgs, human_pose_2D] = human_model.get_img_views(rotations=[30, 120], human_pose_3D=pose, plot_kps=True)
+    else:
+        [out_imgs, human_pose_2D] = human_model.get_img_views(rotations=[30, 120], human_pose_3D=None, plot_kps=False)
```
Done.

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -14,13 +14,16 @@
+     ObjectHypothesisWithPose, Detection3DArray, Pose2D, Detection3D,\
+     BoundingBox3D as BoundingBox3DMsg
+from shape_msgs.msg import Mesh, MeshTriangle
+from geometry_msgs.msg import Point, Pose as Pose3D
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -14,13 +14,16 @@
 
 import numpy as np
 from opendr.engine.data import Image
-from opendr.engine.target import Pose, BoundingBox, BoundingBoxList, Category
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -207,39 +210,132 @@ def from_ros_bounding_box_list(self, ros_detection_2d_array):
+        :return vertices: the vertices of the 3D model
+        :rtype vertices: numpy array (Nx3)
+        :return faces: the faces of the 3D model
+        :rtype faces: numpy array (Nx3)
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_interfaces/srv/Mesh.srv:
```diff
@@ -0,0 +1,7 @@
+sensor_msgs/Image rgb_img
```
Done

charsyme replied:
opendr_interfaces has been removed. Its content is merged with opendr_ros2_messages.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,74 @@
+                                      'human_model_generation/demos/imgs_input/msk/result_0004.jpg'))
+    extract_pose = True
+    rclpy.init()
+    client = Human_model_generation_client()
```
Done

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        Callback that process the input data and publishes to the corresponding topics
+        :param data: input message
+        :type data: sensor_msgs.msg.Image
```
I changed the docstring, in both class methods. I am not sure if they are correct in this case.

charsyme commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,89 @@
+            print("GPU not found. Using CPU instead.")
+            device = 'cpu'
+    except:
+        device = 'cpu'
```
I added argument parsers in both the client and the service.

charsyme commented:
@tsampazk I addressed all the comments. Check those I left unresolved. I ran again the ROS nodes, on foxy version on another machine this time, and it runs smoothly without any errors.
I believe for some reason either the images weren't present in the provided paths or there was an issue ( https://github.com/opendr-eu/opendr/blob/5a1914319ec3163f18bb7e079fc03bbfa144ffe6/src/opendr/simulation/human_model_generation/utilities/PIFu/lib/mesh_util.py#L45 ) which was addressed recently. However, this issue had already been resolved on the branch you reviewed. I hope it is ok now.

tsampazk commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,110 @@
+        :type img_msk: engine.data.Image
+        :param extract_pose: Defines whether to extract the pose of the depicted human or not
+        :type extract_pose: bool
+        :return human_model: The generated human model.
+        :type human_model: opendr.simulation.human_model_generation.utilities.model_3D.Model_3D
+        :return pose: The extracted 3D pose
+        :type pose: engine.target.Pose
```
```suggestion
        :return: A tuple containing the generated human model and the extracted 3D pose
        :rtype: tuple, (opendr.simulation.human_model_generation.utilities.model_3D.Model_3D, engine.target.Pose)
```

tsampazk commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,110 @@
+def main():
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-srv_name", help="The name of the service",
+                        type=str, default="human_model_generation")
+    parser.add_argument("-img_rgb", help="Path for RGB image", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
+                                                                        'human_model_generation/demos/'
+                                                                        'imgs_input/rgb/result_0004.jpg'))
+    parser.add_argument("-img_msk", help="Path for mask image", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
+                                                                        'human_model_generation/demos/'
+                                                                        'imgs_input/msk/result_0004.jpg'))
+    parser.add_argument("-rot_angles", help="Yaw angles for rotating the generated model",
+                        nargs="+", default=['30', '120'])
+    parser.add_argument("-extract_pose", help="Whether to extract pose or not", action='store_true')
+    parser.add_argument("-plot_kps", help="Whether to plot the keypoints of the extracted pose",
+                        action='store_true')
+    parser.add_argument("-out_path", help="Path for outputting the renderings/models", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
```
The argument convention we follow in other nodes is double minus before full arguments and single minus before shortcuts, e.g. `-i` vs `-input`. Just added the missing symbols in the suggestion below, other than that arguments look good!

```suggestion
    parser.add_argument("--srv_name", help="The name of the service",
                        type=str, default="human_model_generation")
    parser.add_argument("--img_rgb", help="Path for RGB image", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
                                                                        'human_model_generation/demos/'
                                                                        'imgs_input/rgb/result_0004.jpg'))
    parser.add_argument("--img_msk", help="Path for mask image", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
                                                                        'human_model_generation/demos/'
                                                                        'imgs_input/msk/result_0004.jpg'))
    parser.add_argument("--rot_angles", help="Yaw angles for rotating the generated model",
                        nargs="+", default=['30', '120'])
    parser.add_argument("--extract_pose", help="Whether to extract pose or not", action='store_true')
    parser.add_argument("--plot_kps", help="Whether to plot the keypoints of the extracted pose",
                        action='store_true')
    parser.add_argument("--out_path", help="Path for outputting the renderings/models", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
```

tsampazk commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_service.py:
```diff
@@ -0,0 +1,109 @@
+def main():
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-device", help="Device to use, either \"cpu\" or \"cuda\", defaults to \"cuda\"",
+                        type=str, default="cuda", choices=["cuda", "cpu"])
+    parser.add_argument("-srv_name", help="The name of the service",
+                        type=str, default="human_model_generation")
+    parser.add_argument("-checkpoint_dir", help="Path to directory for the checkpoints of the method's network",
+                        type=str, default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
+    args = parser.parse_args()
```
Just added the extra minus sign same as the other file arguments.
```suggestion
    parser.add_argument("--device", help="Device to use, either \"cpu\" or \"cuda\", defaults to \"cuda\"",
                        type=str, default="cuda", choices=["cuda", "cpu"])
    parser.add_argument("--srv_name", help="The name of the service",
                        type=str, default="human_model_generation")
    parser.add_argument("--checkpoint_dir", help="Path to directory for the checkpoints of the method's network",
                        type=str, default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
    args = parser.parse_args()
```

tsampazk reviewed (changes requested):
Thank you for the fixes @charsyme ! There are some minor changes that are still needed and we should be good to go.

Regarding the error, i still got it when trying to run the node the `ros2-human-model-generation` branch. I reverted the line to the old version `marching_cubes_lewiner` instead of the current version `marching_cubes` (which also produces a warning `Cannot find reference 'marching_cubes' in '__init__.py | __init__.py' ` in my IDE). With the old version it runs without error, producing 2 images that i attach bellow. One of them seems problematic. However, we don't need to fix this to merge the PR, because it is not related to the ROS2 nodes. Could you investigate further and maybe open an issue to document the error if it is indeed present?

![rendering30](https://user-images.githubusercontent.com/27914645/191955336-35ac35e3-99f1-4476-8133-f938a9989550.jpg)
![rendering120](https://user-images.githubusercontent.com/27914645/191955339-57119238-8c3c-44a6-802a-c8b309ca0024.jpg)

charsyme commented:
> Thank you for the fixes @charsyme ! There are some minor changes that are still needed and we should be good to go.
> 
> Regarding the error, i still got it when trying to run the node the `ros2-human-model-generation` branch. I reverted the line to the old version `marching_cubes_lewiner` instead of the current version `marching_cubes` (which also produces a warning `Cannot find reference 'marching_cubes' in '__init__.py | __init__.py' ` in my IDE). With the old version it runs without error, producing 2 images that i attach bellow. One of them seems problematic. However, we don't need to fix this to merge the PR, because it is not related to the ROS2 nodes. Could you investigate further and maybe open an issue to document the error if it is indeed present?
> 
> ![rendering30](https://user-images.githubusercontent.com/27914645/191955336-35ac35e3-99f1-4476-8133-f938a9989550.jpg) ![rendering120](https://user-images.githubusercontent.com/27914645/191955339-57119238-8c3c-44a6-802a-c8b309ca0024.jpg)

@tsampazk  It is strange because in the corresponding test (which is passed successfully), the number of vertices of the 3D model are counted
https://github.com/opendr-eu/opendr/blob/fa1b2daccd16a5d4e1e5c66994dd5cd254c501be/tests/sources/tools/simulation/human_model_generation/test_human_model_generation.py#L54
which means that the 3D model is generated smoothly during the run of the test. I will check it though in the following days.

charsyme commented on projects/opendr_ws_2/src/opendr_ros2_bridge/opendr_ros2_bridge/bridge.py:
```diff
@@ -243,3 +244,120 @@ def to_ros_face_id(self, category):
+         representing vertices and faces of the 3D model respectively
+        :rtype: tuple (np.array, np.array)
+        :return vertex_colors: the colors of the vertices of the 3D model
+        :rtype vertex_colors: numpy array (Nx3)
```
Indeed I modified it.

charsyme commented:
> > Thank you for the fixes @charsyme ! There are some minor changes that are still needed and we should be good to go.
> > Regarding the error, i still got it when trying to run the node the `ros2-human-model-generation` branch. I reverted the line to the old version `marching_cubes_lewiner` instead of the current version `marching_cubes` (which also produces a warning `Cannot find reference 'marching_cubes' in '__init__.py | __init__.py' ` in my IDE). With the old version it runs without error, producing 2 images that i attach bellow. One of them seems problematic. However, we don't need to fix this to merge the PR, because it is not related to the ROS2 nodes. Could you investigate further and maybe open an issue to document the error if it is indeed present?
> > ![rendering30](https://user-images.githubusercontent.com/27914645/191955336-35ac35e3-99f1-4476-8133-f938a9989550.jpg) ![rendering120](https://user-images.githubusercontent.com/27914645/191955339-57119238-8c3c-44a6-802a-c8b309ca0024.jpg)
> 
> @tsampazk It is strange because in the corresponding test (which get passed successfully), the number of vertices of the 3D model are counted
> 
> https://github.com/opendr-eu/opendr/blob/fa1b2daccd16a5d4e1e5c66994dd5cd254c501be/tests/sources/tools/simulation/human_model_generation/test_human_model_generation.py#L54
> 
> which means that the 3D model is generated smoothly during the run of the test. I will check it though in the following days.

I opened an issue as it is not related to the corresponding ROS2 nodes.

tsampazk commented:
> I opened an issue as it is not related to the corresponding ROS2 nodes.

Thanks @charsyme, feel free to re-request a review if you feel the issues on this PR have been resolved.

tsampazk commented on projects/opendr_ws_2/src/opendr_simulation/opendr_simulation/human_model_generation_client.py:
```diff
@@ -0,0 +1,108 @@
+def main():
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-srv_name", help="The name of the service",
+                        type=str, default="human_model_generation")
+    parser.add_argument("-img_rgb", help="Path for RGB image", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
+                                                                        'human_model_generation/demos/'
+                                                                        'imgs_input/rgb/result_0004.jpg'))
+    parser.add_argument("-img_msk", help="Path for mask image", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
+                                                                        'human_model_generation/demos/'
+                                                                        'imgs_input/msk/result_0004.jpg'))
+    parser.add_argument("-rot_angles", help="Yaw angles for rotating the generated model",
+                        nargs="+", default=['30', '120'])
+    parser.add_argument("-extract_pose", help="Whether to extract pose or not", action='store_true')
+    parser.add_argument("-plot_kps", help="Whether to plot the keypoints of the extracted pose",
+                        action='store_true')
+    parser.add_argument("-out_path", help="Path for outputting the renderings/models", type=str,
+                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
```
Argument names still look like they use single minus instead of double even though the comment and suggestion i made in earlier review is marked as resolved. I am re-adding the suggestion, reasoning can be found in earlier comment.


```suggestion
    parser.add_argument("--srv_name", help="The name of the service",
                        type=str, default="human_model_generation")
    parser.add_argument("--img_rgb", help="Path for RGB image", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
                                                                        'human_model_generation/demos/'
                                                                        'imgs_input/rgb/result_0004.jpg'))
    parser.add_argument("--img_msk", help="Path for mask image", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/python/simulation/'
                                                                        'human_model_generation/demos/'
                                                                        'imgs_input/msk/result_0004.jpg'))
    parser.add_argument("--rot_angles", help="Yaw angles for rotating the generated model",
                        nargs="+", default=['30', '120'])
    parser.add_argument("--extract_pose", help="Whether to extract pose or not", action='store_true')
    parser.add_argument("--plot_kps", help="Whether to plot the keypoints of the extracted pose",
                        action='store_true')
    parser.add_argument("--out_path", help="Path for outputting the renderings/models", type=str,
                        default=os.path.join(os.environ['OPENDR_HOME'], 'projects/opendr_ws_2'))
```

charsyme replied:
Done

tsampazk reviewed (changes requested):
Thanks @charsyme, changes look good apart from the arguments in client node, i left a comment. My guess is that the suggested changes from the earlier review were not applied by mistake.

As a note, to run the nodes successfully i needed to change `marching_cubes` to `marching_cubes_lewiner` again in `mesh_util.py`. Maybe it has something to do with the version of `skimage`? Also noticed that on the pop-up windows, **both images get produced properly**, but when i open the images that are saved to disk `rendering120.jpg` is messed up, relating to #334. I will also comment directly on the issue.

charsyme commented:
> Thanks @charsyme, changes look good apart from the arguments in client node, i left a comment. My guess is that the suggested changes from the earlier review were not applied by mistake.
> 
> As a note, to run the nodes successfully i needed to change `marching_cubes` to `marching_cubes_lewiner` again in `mesh_util.py`. Maybe it has something to do with the version of `skimage`? Also noticed that on the pop-up windows, **both images get produced properly**, but when i open the images that are saved to disk `rendering120.jpg` is messed up, relating to #334. I will also comment directly on the issue.

@tsampazk I applied the changes in the client node. I hope I haven't forgot any requested changes from previous reviews. If that's the case, the PR is ready. Regarding the problem with the images, I set a more strict restriction regarding the version of scikit-image (scikit-image>=0.19), but I haven't tested the results yet. Anyway, there is a separate opened issue, in which this problem will investigated.

tsampazk commented:
Ok @charsyme, thank you very much, i will add new review as soon as the tests complete, but i think the changes are ready to be merged now!
